### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by europ**
* added `.gitignore` from https://github.com/github/gitignore/blob/master/Python.gitignore
    * removed `artifacts` folder
    * removed `artifacts.zip` file
* added unit tests for each file in `atacac/`
* parameter FIX in `atacac/backup.py`
* increased line length for flake8
* added `test` folder to tox variable `sources`
* increased verbosity level for pytest output (option `-v` enabled)
* created custom exception class

### CODE REVIEW

File: test/test_atacac/test_backup.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                        {"b5c3_x": 7, "b5c3_y": 8, "b5c3_z": 7}
+                    ],
+                    "b5_c4": ["x", "y", "z"]
+                }
+            }
+        }
+
+        result = yaml.dump(data, Dumper=backup.Dumper, default_flow_style=False)
+
+        assert result == (
```
Comment (fpob) [inline test/test_atacac/test_backup.py]:
```
assert result == textwrap.dedent(
  """\
  a:
    b1:
  ...
  """
)
```

would be more readable IMO.

File: atacac/_utils.py [hunk 2]
```diff
@@ -59,7 +65,7 @@ def tower_list(asset_type, query):
 
     """
     if asset_type not in ASSET_TYPES:
-        raise Exception('Unsupported asset type \'{}\'!'.format(asset_type))
+        raise Error('Unsupported asset type \'{}\'!'.format(asset_type))
```
Comment (fpob) [inline atacac/_utils.py]:
```suggestion
        raise Error(f"Unsupported asset type '{asset_type}'!")
```

File: atacac/_utils.py [hunk 3]
```diff
@@ -84,7 +90,7 @@ def tower_receive(asset_type, asset_name):
         assets objects
     """
     if asset_type not in ASSET_TYPES:
-        raise Exception('Unsupported asset type \'{}\'!'.format(asset_type))
+        raise Error('Unsupported asset type \'{}\'!'.format(asset_type))
```
Comment (fpob) [inline atacac/_utils.py]:
```suggestion
        raise Error(f"Unsupported asset type '{asset_type}'!")
```

File: atacac/_utils.py [hunk 4]
```diff
@@ -115,5 +121,8 @@ def load_asset(file_path):
 
     Now simply opens file and parses YAML.
     """
-    with open(file_path) as f:
-        return yaml.safe_load(f)
+    try:
+        with open(file_path) as f:
+            return yaml.safe_load(f)
+    except IOError:
+        raise Error("Failed to read content of '{}'".format(file_path))
```
Comment (fpob) [inline atacac/_utils.py]:
```suggestion
        raise Error(f"Failed to read content of '{file_path}'")
```

File: test/test_atacac/test__utils.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+            result = _utils.load_asset(tmpfile.name)
+
+            assert result == yaml.safe_load(file_data)
+
+    else:
+        try:
+            _utils.load_asset("./a/b/c/d/e/f/g/h/i/file")
+        except _utils.Error as e:
+            assert re.match("^Failed to read content of './a/b/c/d/e/f/g/h/i/file'$", str(e))
```
Comment (fpob) [inline test/test_atacac/test__utils.py]:
```suggestion
            assert re.match(r"^Failed to read content of '\./a/b/c/d/e/f/g/h/i/file'$", str(e))
```

File: test/test_atacac/test_backup.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
+import os
+import shutil
+import textwrap
+from unittest import mock
+from test.utils import MockDir
+
+import yaml
+import pytest  # pylint: disable=import-error
```
Comment (fpob) [inline test/test_atacac/test_backup.py]: Why is pylint disabled here and in **all** other files with tests?

File: test/test_atacac/test_backup.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    ]
+)
+@mock.patch("atacac.backup.log")
+@mock.patch("atacac.backup.load_asset")
+@mock.patch("atacac.backup.tower_receive")
+def test_main(mock_tower_receive, mock_load_asset, mock_log, assets_glob):
+    backup_path = "./backup"
+
+    try:
+        shutil.rmtree(backup_path)
```
Comment (fpob) [inline test/test_atacac/test_backup.py]: You should create temp directory instead.

File: test/test_atacac/test_backup.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+import yaml
+import pytest  # pylint: disable=import-error
+from click.testing import CliRunner
+
+from atacac import backup
+
+
+class TestDumper:
+
+    def test_increase_indent(self):
```
Comment (fpob) [inline test/test_atacac/test_backup.py]: Why this test is in class while all other tests are functions?

File: test/test_atacac/test_backup.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+    runner = CliRunner()
+    result = runner.invoke(backup.main)
+
+    assert result.output == ""
+    assert result.exit_code == 0
+    assert result.exception is None
+
+    assert mock_log.call_count == len(good_files) * 3
+    assert mock_load_asset.call_count == len(good_files)
```
Comment (fpob) [inline test/test_atacac/test_backup.py]: You should use `assert_has_calls` here or something which will check calls. Just from number of calls you can't assume `load_asset` was called only for good files.

Comment (fpob) [reply test/test_atacac/test_backup.py]: dtto `     assert len(os.listdir(backup_path)) == len(good_files)` and `     assert len(os.listdir(backup_path)) == len(good_files)`

File: test/test_atacac/test_restore.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+                "VALUE_token",
+                "VALUE_project",
+                artifact_job,
+                "VALUE_pipeline_id",
+                file.name
+            )
+        except Abort:
+            assert (pipeline_status != "success"
+                    or jop_status != "success"
+                    or artifact_name != "artifacts.zip")
```
Comment (fpob) [inline test/test_atacac/test_restore.py:133]: And if `restore.gitlab_download` wasn't raised Abort everthing is OK no matter what `restore.gitlab_download` did?

File: test/test_atacac/test_restore.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+def test_upload(mock_tower_send, mock_log, files):
+    mock_dir = MockDir()
+    mock_dir.add_file(files)
+
+    restore.upload(mock_dir.path)
+
+    yml_files = [file for file in files if file.endswith(".yml")]
+
+    assert mock_log.call_count == len(yml_files) + 1
+    assert mock_tower_send.call_count == len(yml_files)
```
Comment (fpob) [inline test/test_atacac/test_restore.py]: You should check calls. You can't assume that `tower_send` was called with correct arguments.

File: test/test_atacac/test_validators/test_builtin.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+from test.utils import MockDir
+
+import pytest  # pylint: disable=import-error
+import yamale
+from yamale.validators import DefaultValidators
+
+from atacac.validators.builtin import Example, Dict
+
+
+class TestExample:
```
Comment (fpob) [inline test/test_atacac/test_validators/test_builtin.py]: Why class? All other tests are just functions

File: test/test_atacac/test_backup.py (hunk 8 above)
Comment (europ) [reply test/test_atacac/test_backup.py]: I kept the convention:
* Class::method() <=> TestClass::test_method()
* function() <=> test_function()

Comment (europ) [reply test/test_atacac/test_backup.py]: Should we stick to a better one?

File: test/test_atacac/test_validators/test_builtin.py (hunk 12 above)
Comment (europ) [reply test/test_atacac/test_validators/test_builtin.py]: Already mentioned in https://github.com/europ/atacac/pull/11#discussion_r454222298

File: test/test_atacac/test_backup.py (hunk 6 above)
Comment (europ) [reply test/test_atacac/test_backup.py]: `pylint` cannot handle this `import pytest` so I did an inline fix

```
************* Module test.test_atacac.test_validate
test/test_atacac/test_validate.py:7:0: E0401: Unable to import 'pytest' (import-error)
************* Module test.test_atacac.test_backup
test/test_atacac/test_backup.py:7:0: E0401: Unable to import 'pytest' (import-error)
************* Module test.test_atacac.test_synchronize
test/test_atacac/test_synchronize.py:6:0: E0401: Unable to import 'pytest' (import-error)
************* Module test.test_atacac.test___main__
test/test_atacac/test___main__.py:4:0: E0401: Unable to import 'pytest' (import-error)
************* Module test.test_atacac.test_upload
test/test_atacac/test_upload.py:5:0: E0401: Unable to import 'pytest' (import-error)
************* Module test.test_atacac.test_restore
test/test_atacac/test_restore.py:7:0: E0401: Unable to import 'pytest' (import-error)
************* Module test.test_atacac.test_differentiate
test/test_atacac/test_differentiate.py:6:0: E0401: Unable to import 'pytest' (import-error)
************* Module test.test_atacac.test__utils
test/test_atacac/test__utils.py:7:0: E0401: Unable to import 'pytest' (import-error)
************* Module test.test_atacac.test_validators.test___init__
test/test_atacac/test_validators/test___init__.py:6:0: E0401: Unable to import 'pytest' (import-error)
************* Module test.test_atacac.test_validators.test_builtin
test/test_atacac/test_validators/test_builtin.py:6:0: E0401: Unable to import 'pytest' (import-error)

```

Comment (fpob) [reply test/test_atacac/test_backup.py]: `pylint` can handle it and is showing valid warning about missing package `pytest`. You forgot to install it in tox env.

```diff
diff --git a/tox.ini b/tox.ini
index f856b20..6481021 100644
--- a/tox.ini
+++ b/tox.ini
@@ -30,7 +30,9 @@ commands = mypy --ignore-missing-imports {[VARS]sources}
 
 
 [testenv:pylint]
-deps = pylint>=2.5.3
+deps =
+    pylint>=2.5.3
+    pytest>=5.4.3
 commands = pylint -E {[VARS]sources}
```

File: test/test_atacac/test_backup.py (hunk 8 above)
Comment (fpob) [reply test/test_atacac/test_backup.py]: IMHO weird convention :confused: I prefer consistency

Comment (europ) [reply test/test_atacac/test_backup.py]: Same case for https://github.com/europ/atacac/pull/11#discussion_r454239087

What is your suggestion for [`Example::_is_valid()`](https://github.com/europ/atacac/blob/master/atacac/validators/builtin.py#L7) and [`Dict::_is_valid()`](https://github.com/europ/atacac/blob/master/atacac/validators/builtin.py#L14)?

Currently, it is named as `TestExample::test__is_valid()` and `TestDict::test__is_valid()` while it is in a test class.

Can it be a function such as `test_Example__is_valid()` and `test_Dict__is_valid()` ... following pattern of `test_<class>_<method>` ?

Comment (fpob) [reply test/test_atacac/test_backup.py]: Name doesn't really matter if it is descriptive. What about `test_example_validation`, `test_dict_validation`, `test_yaml_dumper_indentation`?

Comment (europ) [reply test/test_atacac/test_backup.py]: Alright. I will add it.

File: test/test_atacac/test__utils.py [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+@pytest.mark.parametrize(
+    "valid",
+    [
+        pytest.param(True, id="valid file path"),
+        pytest.param(False, id="invalid file path")
+    ]
+)
+def test_load_asset(valid):
+    if valid:
```
Comment (fpob) [inline test/test_atacac/test__utils.py]: You could split this test case to 2 functions and get rid of this `if`. It is better to keep tests as simple as possible.

**Review by fpob — APPROVED** [review]

**Pull request merged**